helper/maputil/syncmap.go

```diff
@@ -0,0 +1,83 @@
+package maputil
+
+import "sync"
+
+// SyncMap implements a map similar to sync.Map, but with generics and with an equality
+// in the values specified by an "ID()" method.
+type SyncMap[K comparable, V IDer] struct {
+	// lock is used to synchronize access to the map
+	lock sync.RWMutex
+	// data holds the actual data
+	data map[K]V
+}
+
+// NewSyncMap returns a new, empty SyncMap.
+func NewSyncMap[K comparable, V IDer]() *SyncMap[K, V] {
+	return &SyncMap[K, V]{}
+}
+
+// Get returns the value for the given key.
+func (m *SyncMap[K, V]) Get(k K) V {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.data[k]
+}
+
+// Pop deletes and returns the value for the given key, if it exists.
+func (m *SyncMap[K, V]) Pop(k K) V {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	v, ok := m.data[k]
+	if ok {
+		delete(m.data, k)
+	}
+	return v
+}
+
+// PopIfEqual deletes and returns the value for the given key, if it exists
+// and only if the ID is equal to the provided string.
+func (m *SyncMap[K, V]) PopIfEqual(k K, id string) *V {
```

Comment:
Any reason this shouldn't directly return the value rather than a pointer to it, similar to `Pop`?

```suggestion
func (m *SyncMap[K, V]) PopIfEqual(k K, id string) V {
```

Reply:
For some reason my IDE wouldn't let me return `nil` here unless it was declared as `*V` (though I'm pretty sure `nil` is a valid return value for an interface like `V`). I'll try again. :)

Reply:
Oh I see now, `connPopIfEqual` returned `nil`. The only constraint on `V` is that it has an `ID()` method, so it seems legit to not be able to return `nil` as it could be a pass-by-value struct, not a pointer.

I think the solution is probably to `return v`, and that will be the zero-value of whatever `V` is, i.e. `nil` for our only existing use-case where it's a pointer. As long as it's documented as part of the API, I think those semantics are reasonable for non-pointers too, and it makes the function consistent with `Pop`.

Reply:
I suppose another reasonable solution might be to edit the function signature slightly to:

```go
func (m *SyncMap[K, V]) PopIfEqual(k K, id string) (V, bool) {
```